collaboration/extending-kyma-cli.md

```diff
@@ -0,0 +1,230 @@
+
+## Architecture
+
+  We should follow the modular approach with each component having its own library. This library can be used inside the kyma cli for acessing various components. The library contains all the implementation. The kyma cli should invoke this library. We can take inspiration from hydroform and implement it in similar way. Currently we have following modules in kyma:
+
+  1. Lambdas
+  2. Service Catalog
+  3. Events
+  4. Application Connector
+  5. APIs
+
+## Command structure
+
+  The command structure used by cloud providers is following:
+
+  ```bash
+  <cliName> <functionality> <action>
+  #example aws s3 ls
+  ```
+
+  The current kyma-cli also uses the same pattern eg.
+
+  ```bash
+  kyma provision
+  or
+  kyma install
+  ```
+
+### Lambdas
+
+For the serverless component we should start with supporting following commands
+
+#### Local workspace setup
+
+1. Creation of project locally
+  `kyma lambda init <lambda-name> --runtime <runtime-name> -n <namespace>` or `kyma lambda init <lambda-name> --runtime <runtime-name> -n <namespace> -p <path>`
+  This should create directoy with following content based on the runtime:
+
+      ```bash
+        <lambda-name>
+        ├── .vscode
+        │   └── launch.json
+        ├── deployment
+        │   └── deployment.yaml
+        ├── local
+        │   └── index.js
+        └── code
+          ├── config.yaml
+          ├── handler.js
+          └── package.json
+      ```
+
+    > When path `-p` is passed then it should create in desitnation path else it use the workspace set in the yaml file.
+
+    * launch.json consists of settings for the debugging
+    * deployment.yaml consists of the yaml that has been deployed on the kyma cluster.
+    * `index.js` consists of the server code with handler pointing to `handler.js`
+    * config.yaml can be used as info file with current configurations like:
+      - name
+      - namespace
+      - events subscribed
+      - environment variables
+      - Api name to expose with
+      - replicas
+      - min/max memory
+      - min/max cpu
+      - runtime
+      - debug command (to be used with telepresence. This can help if we want to have different runtimes)
+    * handler.js where the developer can write his logic to test
+    * package.json file with dependencies.
+
+    >The templates generated should be placed a directory in the kyma cli. In the future if we are supporting more runtimes we can move it to separate repo.
+
+2. Testing/debugging it locally
+
+We can introduce commands to test the lambda locally and also if the developer wants to debug the code. This is useful in case when we have some event payloads that are not easy to recreate.
+
+`kyma lambda debug <lambda-name>`
+
+> We can use telepresence for local debugging. We can either use --swap-deployment or --new-deployment option
+
+#### Working with deployed kyma cluster
```

Comment:
Will be good to have a separate listing of the commands supporting the config.yaml to understand what plays in to the main lambda project scenario. So for event binding I would also pass the project folder as argument?

It is nice to see the overall picture on what might be supported long-term but we should focus more on the lambda project scenario. What commands a user will use e2e when having such lambda project? WHat are the typical flows and with that the command chains?

Reply:
updated.